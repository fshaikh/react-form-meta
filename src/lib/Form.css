form {
    width: 50%;
    margin: 10px;
}
legend {
    border-bottom: 1px solid #e5e5e5;
}

.validation-error{
    color: red;
}