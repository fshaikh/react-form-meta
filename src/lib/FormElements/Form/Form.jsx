/**
 * React specific imports
 */
import React from 'react';
import PropTypes from 'prop-types';

/**
 * Lib specific imports
 */
import 'bootstrap/dist/css/bootstrap.min.css';
import Title from '../Title';
import Description from '../Description';

const Form = (props) => {
    const {
        title,
        description,
        rootIdPrefix,
        id,
        name,
        className,
        autoComplete
    } = props.formProperties;

    return (
        <form 
            id = {id}
            name = {name}
            className = {className}
            autoComplete = {autoComplete.toString()}
            
            >
            <div className="form-group">
                <fieldset>
                    <Title title={title} />
                    <Description description={description} />
                    <div>
                        {props.children}
                    </div>
                </fieldset>
            </div>
            {props.actionButtons}
        </form>
    )
};

Form.propTypes = {
    formProperties: PropTypes.object,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    actionButtons: PropTypes.element.isRequired
};

export default Form;