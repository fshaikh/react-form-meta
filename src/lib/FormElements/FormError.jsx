import React from 'react';
import PropTypes from 'prop-types';

/**
 * Component for defining error boundary for Form component
 */
export default class FormError extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            hasError: false,
            error: null,
            errorInfo: null
        };
    }

    /**
     * 
     * @param {*} error - error that has been thrown.
     * @param {*object} info - object with componentStack key. The property has information about component stack during thrown error.
     */
    componentDidCatch(error, errorInfo) {
        this.setState({hasError: true, error: error, errorInfo: errorInfo});
        this.props.onRenderError(error, errorInfo);
    }

    render() {
        if(this.state.hasError) {
            return (
                <div>
                    <h2>Something went wrong.</h2>
                    <details style={{ whiteSpace: 'pre-wrap' }}>
                        {this.state.error && this.state.error.toString()}
                        <br />
                        {this.state.errorInfo.componentStack}
                    </details>
                </div>
            );
        }
        return this.props.children;
    }
}

FormError.propTypes = {
    children: PropTypes.node.isRequired,
    onRenderError: PropTypes.func
};