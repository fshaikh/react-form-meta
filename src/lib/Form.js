import React from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import './Form.css';

import FieldHOC from './Fields/FieldHOC';
import getFieldControl from './FieldControlFactory';
import Title from './FormElements/Title';
import Description from './FormElements/Description';
import SubmitButton from './FormElements/SubmitButton';
import getValidator from './validators/validatorFactory';

export default class Form extends React.Component {
    constructor(props) {
        super(props);
        // Holds form data. If client has passed formData object use it else initialize to empty object
        this.formData = this.props.formData || {};
        
        // Hold the form schema
        this.schema = this.props.schema;

        // Hold the field validators
        this.validators = [];

        this.state = {
            showError: false,
            validationMessages: {}
        };

        // Bind the event handlers
        this.onSubmit = this.onSubmit.bind(this);
        this.onChange = this.onChange.bind(this);
    }

    render() {
        return (
            <form>
                <div className="form-group">
                    <fieldset>
                        <Title title={this.schema.title} />
                        <Description description={this.schema.description} />
                        <div>
                            {this.getFormBody()}
                        </div>
                    </fieldset>
                </div>
                <SubmitButton onClick = {(e) => {this.onSubmit(e)}} />
            </form>
        )
    }

    onSubmit(e) {
        e.preventDefault();
        // perform form validation based on each field validator
        this.validate();
        this.props.onSubmit(this.formData)
    }

    onChange(event, name, value) {
        this.formData[name] = value != null ? value : event.target.value;
        this.validate();
    }

    getFormBody() {
        return this.schema.properties.map((fieldControl) => {
            const Field = getFieldControl(fieldControl.type);
            this.setValidators(fieldControl);
            return <FieldHOC field={fieldControl}
                             key={fieldControl.name}
                             showError={this.state.showError}
                             errorMessages={this.getValidationMessage(fieldControl.name)}>
                        {this.getField(Field, fieldControl)}
                   </FieldHOC>
        });
    }

    getField(Field, fieldControl) {
        return <Field field={ fieldControl }
                        onChange = {this.onChange}
                        value={this.formData[fieldControl.name] || ''}
                />
    }

    

    setValidators(fieldControl) {
        this.validators.push({
                                field: fieldControl.name,
                                fieldTitle: fieldControl.label,
                                validators: fieldControl.validators
                            });
    }

    validate() {
        var validationMessages = {};
        this.validators.forEach((fieldValidator) => {
            const field = fieldValidator.field;
            const fieldTitle = fieldValidator.fieldTitle;
            const validators = fieldValidator.validators || [];
            // get the field value from the form data
            const fieldValue = this.formData[field];
            validationMessages[field] = [];
            validators.forEach((validator) => {
                // call the validator
                const args = [field, fieldValue, fieldTitle, ...validator.args];
                const errorMessage = getValidator(validator.name)(...args);
                if(!errorMessage.isValid){
                    validationMessages[field].push(errorMessage);
                }
                
            });

        });
        if(Object.keys(validationMessages).length > 0) {
            this.setState((prevState, props) => {
                return {
                    showError: true,
                    validationMessages: validationMessages
                }
            });
        }
    }

    getValidationMessage(field) {
        const messages = this.state.validationMessages[field];
        if(messages == null) {
            return [];
        }
        return messages.map((message) => message.errorMessage);
    }
}