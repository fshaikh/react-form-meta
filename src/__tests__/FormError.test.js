/**
 * Test suite for FormError component which acts as a error boundary for Form React component
 */
import React from 'react';
import renderer from 'react-test-renderer';
import FormError from '../lib/FormElements/FormError';
import Form from '../lib/Form';

describe('Form Error component',() => {
    it('should render child view when no error', () => {
        const schema = {
            title: 'Form',
            properties:[]
        };
        const tree = renderer.create(<FormError onRenderError={()=>{} }>
                                        <Form schema={schema} />
                                     </FormError>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should render error view when child component throws',() => {
        
    });
});