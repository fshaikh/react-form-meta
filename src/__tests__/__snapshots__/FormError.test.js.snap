// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Error component should render child view when no error 1`] = `
<form
  onChange={[Function]}
  onFocus={undefined}
>
  <div
    className="form-group"
  >
    <fieldset>
      <legend>
        Form
      </legend>
      <p
        className="field-description"
      />
      <div />
    </fieldset>
  </div>
  <div>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Submit
    </button>
  </div>
</form>
`;
